pkgs/build-support/gradle/update-deps.nix

```diff
@@ -0,0 +1,168 @@
+{ lib
+, pkgs
+, stdenv
+, writeScript
+, writeShellScript
+, nix
+, gradle
+, openssl
+, mitm-cache
+, jdk
+, nsjail
+, path
+, python3
+, coreutils
+, bash
+, procps
+}:
+
+# path to the derivation attr (usually the package name)
+{ attrPath ? pname
+# pname (attrPath synonym)
+, pname ? ""
+# the Nix code to get the derivation (this will be put in Bash double quotes)
+# this is relevant for downstream (non-nixpkgs) users
+, code ? "with import ./. { }; ($attrPath)"
+# unlike gradleFlags in normal packages, these flags are escaped so they are allowed to contain spaces
+# also, gradleFlags from configurePhase are used by default
+, gradleFlags ? []
+# attribute overrides for the "source" derivation
+, attrOverrides ? {}
+# run before fetching the dependencies (in sandbox)
+, preBuild ? ""
+# run before starting the sandbox for fetching dependencies
+, preBuildNoSandbox ? ""
+# the gradle commands for fetching the necessary dependencies
+, gradleCommands ? "gradle nixDownloadDeps"
+# run after fetching the dependencies (in sandbox)
+, postBuild ? ""
+# run after the sandbox for fetching dependencies has stopped
+, postBuildNoSandbox ? ""
+, mitmCachePort ? null
+# deps path (relative to the package directory), or absolute
+, depsPath ? "deps.json"
+# extra packages to make available in the sandbox
+, availablePackages ? [ ]
+, useNsjail ? stdenv.isLinux
+}:
+
+let
+  mitmCachePortBash =
+    if mitmCachePort != null then "${toString mitmCachePort}"
+    else "$(${python3.pkgs.ephemeral-port-reserve}/bin/ephemeral-port-reserve 127.0.0.1)";
+  keep = [ "MITM_CACHE_PORT" "MITM_CACHE_KEYSTORE" "MITM_CACHE_KS_PWD" "MITM_CACHE_CA" ];
+  # FIXME: storePath isn't allowed in pure evaluation mode. Is there a better way of pinning the dependencies?
+  availablePackages' = map (drv: ''(builtins.storePath "${drv}")'') (availablePackages ++ [ bash coreutils ]);
+  gradleScript = writeScript "gradle-commands.sh" ''
+    #!/usr/bin/env nix-shell
+    #!nix-shell --pure ${lib.concatMapStringsSep " " (x: "--keep ${x}") keep} -i bash -p ${lib.concatStringsSep " " availablePackages'}
```

Comment:
Could you use a normal shell as the interpreter here? That way you don't have to use anything like `builtins.storePath`, and the script won't have to rely on `NIX_PATH`.

Reply:
I'd love to, but it's important to set the relevant `PATH`, `PKG_CONFIG_PATH`, etc variables, how do I do that?

Reply:
Unless there is an easier approach, I suggest using `writeShellApplication` as per my [previous comment](https://github.com/NixOS/nixpkgs/pull/272380#discussion_r1658619890), since you can just add dependencies to `runtimeInputs`:

<details>

<summary> gradleScript.patch </summary>

```diff
diff --git a/pkgs/build-support/gradle/update-deps.nix b/pkgs/build-support/gradle/update-deps.nix
index 46a1a4e6c7b9..fa41da586ce2 100644
--- a/pkgs/build-support/gradle/update-deps.nix
+++ b/pkgs/build-support/gradle/update-deps.nix
@@ -3,6 +3,7 @@
 , stdenv
 , writeScript
 , writeShellScript
+, writeShellApplication
 , nix
 , gradle
 , openssl
@@ -51,47 +52,50 @@ let
     if mitmCachePort != null then "${toString mitmCachePort}"
     else "$(${python3.pkgs.ephemeral-port-reserve}/bin/ephemeral-port-reserve 127.0.0.1)";
   keep = [ "MITM_CACHE_PORT" "MITM_CACHE_KEYSTORE" "MITM_CACHE_KS_PWD" "MITM_CACHE_CA" ];
-  # FIXME: storePath isn't allowed in pure evaluation mode. Is there a better way of pinning the dependencies?
-  availablePackages' = map (drv: ''(builtins.storePath "${drv}")'') (availablePackages ++ [ bash coreutils ]);
-  gradleScript = writeScript "gradle-commands.sh" ''
-    #!/usr/bin/env nix-shell
-    #!nix-shell --pure ${lib.concatMapStringsSep " " (x: "--keep ${x}") keep} -i bash -p ${lib.concatStringsSep " " availablePackages'}
-    set -eo pipefail
-    source="$1"
-    gradleArgs=(
-      ${lib.escapeShellArgs gradleFlags}
-      --no-daemon
-      --console plain
-      --init-script ${./init-deps.gradle}
-      -Dhttp.proxyHost="127.0.0.1"
-      -Dhttps.proxyHost="127.0.0.1"
-      -Dhttp.proxyPort="$MITM_CACHE_PORT"
-      -Dhttps.proxyPort="$MITM_CACHE_PORT"
-      -Djavax.net.ssl.trustStore="$MITM_CACHE_KEYSTORE"
-      -Djavax.net.ssl.trustStorePassword="$MITM_CACHE_KS_PWD"
-    )
-    unset MITM_CACHE_KEYSTORE
-    unset MITM_CACHE_KS_PWD
-    function gradle() {
-      ${gradle}/bin/gradle $gradleFlags "''${gradleFlagsArray[@]}" "''${gradleArgs[@]}" "$@"
-    }
-    export GRADLE_USER_HOME="$(mktemp -d)"
-    export TERM=dumb
-    trap "rm -rf '$GRADLE_USER_HOME'" SIGINT SIGTERM ERR EXIT
-    export http_proxy="127.0.0.1:$MITM_CACHE_PORT"
-    export https_proxy="127.0.0.1:$MITM_CACHE_PORT"
-    export NIX_SSL_CERT_FILE="$MITM_CACHE_CA"
-    export SSL_CERT_FILE="$MITM_CACHE_CA"
-    SRC="$(mktemp -d)"
-    trap "rm -rf '$SRC'" SIGINT SIGTERM ERR EXIT
-    cp -r "$source/src" "$SRC/"
-    chmod -R +w "$SRC/src"
-    cd "$SRC/src"
-    source "$source/script"
-    ${preBuild}
-    ${gradleCommands}
-    ${postBuild}
-  '';
+  gradleScript = writeShellApplication {
+    name = "gradle-commands.sh";
+    runtimeInputs = [ coreutils ] ++ availablePackages;
+    text = ''
+      set +o nounset
+      source="$1"
+      gradleArgs=(
+        ${lib.escapeShellArgs gradleFlags}
+        --no-daemon
+        --console plain
+        --init-script ${./init-deps.gradle}
+        -Dhttp.proxyHost="127.0.0.1"
+        -Dhttps.proxyHost="127.0.0.1"
+        -Dhttp.proxyPort="$MITM_CACHE_PORT"
+        -Dhttps.proxyPort="$MITM_CACHE_PORT"
+        -Djavax.net.ssl.trustStore="$MITM_CACHE_KEYSTORE"
+        -Djavax.net.ssl.trustStorePassword="$MITM_CACHE_KS_PWD"
+      )
+      unset MITM_CACHE_KEYSTORE
+      unset MITM_CACHE_KS_PWD
+      function gradle() {
+        # shellcheck disable=SC2154
+        ${gradle}/bin/gradle "''${gradleFlags}" "''${gradleFlagsArray[@]}" "''${gradleArgs[@]}" "$@"
+      }
+      # shellcheck disable=SC2155
+      export GRADLE_USER_HOME="$(mktemp -d)"
+      export TERM=dumb
+      trap 'rm -rf "$GRADLE_USER_HOME"' SIGINT SIGTERM ERR EXIT
+      export http_proxy="127.0.0.1:$MITM_CACHE_PORT"
+      export https_proxy="127.0.0.1:$MITM_CACHE_PORT"
+      export NIX_SSL_CERT_FILE="$MITM_CACHE_CA"
+      export SSL_CERT_FILE="$MITM_CACHE_CA"
+      SRC="$(mktemp -d)"
+      trap 'rm -rf "$SRC"' SIGINT SIGTERM ERR EXIT
+      cp -r "$source/src" "$SRC/"
+      chmod -R +w "$SRC/src"
+      cd "$SRC/src"
+      # shellcheck disable=SC1091
+      source "$source/script"
+      ${preBuild}
+      ${gradleCommands}
+      ${postBuild}
+    '';
+  };
 in
 writeShellScript "fetch-deps.sh" ''
   set -eo pipefail
@@ -146,10 +150,9 @@ writeShellScript "fetch-deps.sh" ''
     # nsjail isn't necessary, it's only used to prevent messy build scripts from touching ~
     ${nsjail}/bin/nsjail \
       -N --keep_caps --chroot / --disable_rlimits \
-      -T /home -T /root -B /tmp -R ${toString path} \
+      -B /tmp \
       ${lib.escapeShellArgs (map (x: "-E${x}") keep)} \
-      -E NIX_PATH=nixpkgs=${toString path} \
-      ${gradleScript} "$source"
+      "${gradleScript}/bin/gradle-commands.sh" "$source"
   else
     NIX_PATH=nixpkgs=${toString path} ${gradleScript} "$source"
   fi
-- 
2.45.1
```

</details>

Also, you only really need to mount `/tmp` in the sandbox.

Reply:
Where do those vars come from currently? I don't see any mention of `PKG_CONFIG_PATH` for instance.

Also note that you can always turn attrs into shell script env vars using `lib.toShellVars attrs`.

Reply:
I've successfully done this locally, I'll push the changes tomorrow

Reply:
Done, the script is somewhat rewritten now. It still relies on nix-shell (it basically has to, even if I don't use nix-shell directly I have to reimplement the same logic), but it doesn't use builtins.storePath anymore.

Reply:
Ah, I see why that's needed.
Perhaps instead of passing an attribute to `nix-shell`, you could pass `pkg.drvPath` to it? That way you don't need to set up any evaluation related things in this sandbox. Make that `builtins.unsafeDiscardOutputDependency pkg.drvPath` so that it doesn't have to first build everything; an old Nix quirk we can't just get rid of easily. It's not unsafe, unless you expected to find realised outputs in those `.drv` files, which would be a bit silly.

Ultimately I want the `nix-shell` behavior you need to be part of Nixpkgs. `nix-shell` is mostly a Nixpkgs-only affair anyway. That will get rid of scope creep in Nix, give Nixpkgs and users a bit more freedom, and make mostly obsolete the ugliness I just described about `drvPath`s.
More context: https://github.com/NixOS/nix/issues/7501

For now I think you could just pass something like `builtins.unsafeDiscardOutputDependency pkg.drvPath` to `nix-shell`, where `pkg` is the "derivation" for the shell, which you can evaluate outside of the sandbox.